### ISSUE TITLE
Suggestion: option to run a shell command (or plugin) on job failure?

### ISSUE THREAD

**Issue by aldanor** [jhuckaby/Cronicle#39]
First, thanks @jhuckaby for a wonderful library and reacting so fast to the issues :)

I wanted to suggest an extremely useful feature which seems to be reasonably simple to implement as well - basically, an option to run a shell command (or any 'Plugin' action that can be currently scheduled, for that matter) on job failure. (Of course, it's kind of possible right now, by baking errror handling directly into each task's script, but might be neater if it was separate.)

This would be immensely useful, e.g. in the case of error reporting systems like sentry.io  that provide command line interfaces to report errors to central tracker. However, it could also be some user-made shell script that reacts to an error. In this case, it would also help if the details were exported to the environment (even it's just one json blob dumped to a variable since you can always split it with jq; but could also be separate variables), so you could do stuf like this and get it logged centrally:

```sh
echo $CRONICLE_EVENT | jq '.category, .id | read -d '\n' CATEGORY EVENT_ID
sentry-cli --api-key 1234 --auth-token 5678 -e category:$CATEGORY -e event_id:$EVENT_ID
```

By the way, speaking of environment variables, it might help to always have something like `CRONICLE_EVENT` logged to environment so that jobs can extract scheduler data from them (even if they haven't failed) - for example for logging purposes. If the task log is available as a file somewhere, could potentially share the filename in an environment variable as well. 

And finally, it may also make sense to have a 'Plugin' action when a task is completed as well (wherever it failed or not, unconditionally) -- for example you may want to upload the task details or log (if it's available) somewhere.

Thanks! (and apologies if I missed some already existing functionality from the docs)

**Comment by jhuckaby** [comment]
Hey @aldanor,

Thanks!  I think some of what you are asking for already exists, so let me point you to a few spots in the docs, just to make sure.

So there's this Cronicle feature called [Chain Reaction](https://github.com/jhuckaby/Cronicle#chain-reaction) which allows you to configure any event to run after a job completes.  This gets interesting when you consider that it is configurable dynamically inside of jobs using the JSON API over STDOUT: [Chain Reaction Control](https://github.com/jhuckaby/Cronicle#chain-reaction-control).  Meaning, your own event code can trigger any other event to run, based on any outcome you want.  You can even do this from within a shell script, if you are using the [Shell Plugin](https://github.com/jhuckaby/Cronicle#built-in-shell-plugin).  Just echo some simple JSON and exit. 
 Example:

```sh
echo '{"chain":"e29bf12db"}'
exit 0;
```

This would complete the current job, and spawn event with ID `e29bf12db`.  The idea here is, you can detect your own error condition, and then run this shell code at the appropriate time.  So that way you can spawn a particular event on error, and/or a different one on success, etc.  A more complete shell script example could be:

```sh
CMD="/path/to/my/shell-script.sh --somearg --otherarg";

if $CMD ; then
	echo "Success!";
	exit 0;
else
	echo "ERROR!  Running special chain reaction event.";
	echo '{"chain":"e29bf12db"}';
	exit 0;
fi
```

Cronicle interprets line-delimited JSON as commands, and you can also emit other text to STDOUT, which is logged but ignored.  You can also add custom data to accompany the chained event, using the [Chain Data](https://github.com/jhuckaby/Cronicle#chain-data) feature.  This way you can pass along an error code, error message, or whatever you want to your chained event.  Example:

```sh
echo '{ "chain": "e29bf12db", "chain_data": { "error_code": 421, "error_msg": "The thing failed because reasons." } }'
exit 0;
```

Your chained event will then have access to the `chain_data` object inside the JSON blob that is sent to its STDIN at launch.

Note that the Chain Reaction event only fires if the first job "succeeds" with an error code of 0, hence in all the shell script examples above I am calling `exit 0;`.  If you actually want need the first job to fail, for things like error e-mail notification, note that you can [dynamically change the e-mail notification settings](https://github.com/jhuckaby/Cronicle#changing-notification-settings) inside your job code via JSON.

Another way to accomplish error notifications is to use the [Web Hook](https://github.com/jhuckaby/Cronicle#event-web-hook) system.  Simply provide a URL, and it will be pinged with an HTTP POST on every event completion, with detailed information sent via JSON.  You can see the actual error code, message, and all the other details.  Inside of your web hook code you can then call an external service such as sentry.io.  You can also make an HTTP GET call back to Cronicle to fire off another event, using the [run_event](https://github.com/jhuckaby/Cronicle#run_event) API.

You mentioned some concerns about environment variables.  Note that Cronicle populates the following environment variables for every job:

| Variable | Description |
|----------|-------------|
| `CRONICLE` | The current Cronicle version, e.g. `1.0.0`. |
| `JOB_ID` | The unique alphanumeric ID assigned to the job. |
| `JOB_LOG` | The filesystem path to the job log file, which you can append to if you want.  However, STDOUT or STDERR are both piped to the log already. |
| `JOB_NOW` | The Epoch timestamp of the job time (may be in the past if re-running a missed job). |

In addition, any of the event's [Plugin Parameters](https://github.com/jhuckaby/Cronicle#plugin-parameters) are also passed as environment variables.

See: [Environment Variables](https://github.com/jhuckaby/Cronicle#environment-variables).

Please let me know if these features satisfy your needs, or if you think we really need to introduce a new "run event on error" feature.

Thanks!

**Comment by aldanor** [comment]
Thanks a lot for the detailed reply! I may have overlooked a few chain-reaction-related details, indeed - so thanks for pointing out. I'll try to note some of my concerns and thoughts.

First, after a bit of thinking, the Chain Reaction mechanism is indeed a bit more generic than simply adding an option to call a 'run on error', since you can reuse it in many tasks (that being said, one could argue that you could also reuse a shared shell script in many tasks, given it is present on all nodes). My main concern about it is that it's basically a strict 'on success' trigger:

> Note that the Chain Reaction event only fires if the first job "succeeds" with an error code of 0

This to me seems like the biggest restriction for the chain reaction events. If I wanted to trigger an error handling event (e.g., log the error to the central scheduler like described above), then the parent event must complete successfully -- which seems somewhat wrong, if only because it would then be listed with a green checkmark in the list of completed events, so it would be hard to visually discern what actually failed and what didn't.

So, 'chain-reaction' is actually more of a 'success-reaction'; I guess, one potential workaround to deal with both sides of the problem would be adding a separate 'fail-reaction' trigger, which would only run if the parent event failed. Hypothetically, 'success-reaction' and 'fail-reaction' could even point to the same child event ID without any problem (although it would probably make sense to at least include the exit code of the parent process in JSON that's passed through). Whether it should also read line commands from STDOUT, or maybe STDERR or something, I'm not entirely sure.

With both 'chain-reaction' and 'fail-reaction' pointing to the same post-event handler (and with exit code being passed to the child event in some fashion), your example above could then become simply

```sh
/path/to/my/shell-script.sh --somearg --otherarg
echo '{"chain":"e29bf12db"}'
```

Just thinking about it, another option to extend chain-reaction triggers is to let the user select whether it triggers on success (current behaviour), or fail, or both. This, however, would be more limited as it would eliminate the possibility of using a shared error-only handler for multiple events.

> Another way to accomplish error notifications is to use the Web Hook system.

Yep, that's the first thing I looked at. However, it wouldn't work too well with existing REST APIs that have their own view on how the payload should look like, unless you're willing to maintain a little service that would transform the payloads. If it was possible to provide it with a user-defined JSON template(s) (handlebars template or something) it could probably work, e.g. something like

```json
{"message": "[{{category}}] {{event_title}}", "level": "error", "id": "{{id}}"}
```

It's just a quite random thought though.

> Cronicle interprets line-delimited JSON as commands

A very minor concern, but I actually remember some jobs I had to maintain that emitted JSON as part of their natural logging process. It would be hypothetically neat to be able to make those work as well (e.g., by looking for a user-provided prefix before JSON line commands, defaulting to the empty string).

> Note that Cronicle populates the following environment variables for every job

It would be definitely handy for shell scripts if there were more environment variables available so you wouldn't have to parse JSON from command line (which is doable with jq, but nevertheless). Most if not all build servers like Jenkins, Bamboo, Travis set a myriad of namespaced environment variables that come quite useful. Plus, as I've mentioned above, if chain reaction (or, rather, 'fail-reaction') is allowed for failed events, it would be definitely handy to have the exit code somewhere easily accessible because you will be most likely to check it the first thing.

Thanks!

**Comment by jhuckaby** [comment]
Thanks for the detailed suggestions.  I agree with all of them.  I think I see at least 4 distinct feature requests here:

- Extend chain reaction system to be able to select a chain success and/or a chain failure event.
- Chained event should automatically receive additional details (both in JSON and env vars) from original job, including error code and error message.
- New Shell Plugin option checkbox to disable interpreting JSON commands in STDOUT, in case shell script naturally outputs its own unrelated JSON, which could muck things up.
- Populate more environment variables for all jobs, potentially for every item in the [Event Data](https://github.com/jhuckaby/Cronicle#event-data-format), using some kind of namespace / prefix, e.g. `JOB_`.

Sound good?

**Comment by aldanor** [comment]
@jhuckaby Yes, you've summarized it perfectly well, thanks! Would absolutely love to see it implemented. If it sounds like too much, please let me know if I can help -- would take me a while to read through the full codebase first though.

**Comment by jhuckaby** [comment]
Thanks!  You are of course welcome to submit a PR, but I think I'm sure I can manage these features.  They are all fairly straightforward :)

**Comment by aldanor** [comment]
Indeed, I think so too. I'd leave it to you then of course, thanks! :)

**Comment by jhuckaby** [comment]
Fixed in [v0.7.5](https://github.com/jhuckaby/Cronicle/releases/tag/v0.7.5).  Thanks!

**Comment by aldanor** [comment]
This is massively helpful, I haven't had a chance to play with it in full, but am definitely planning to, there's definitely many new options now; thanks!
